### DISCUSSION TITLE
How to organize my models to make them import faster?

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#9591] [Usage Questions] [answered]
I have about 3000 models reflected from database, and I create a package which provide some data mining functions based on about 100 of them.

My problem is, import my module takes more than 5 second, and I want to provide all 3000 models to my users so they can extend my functions on demand.

Any one have suggestions how can I organize my models to make my package import faster and keep all models in it.

I found ORM class is **a bit** slower than simple python class. But 3000 **a bit** is a lot amount time.

This is my import profile generated by `python -X importtime tests/profile_import.py > import2.log 2>&1`

ORM:
```
...
import time:      4999 |       5349 |           sqlalchemy.orm.events
import time:       252 |        252 |           sqlalchemy.orm.dependency
import time:       851 |      36544 |         sqlalchemy.orm
import time:       426 |      59059 |       mymodel.utils
import time:       891 |     852126 |     mymodel.client.base
import time:   1900561 |    1900561 |       mymodel.client.db1.models
import time:       261 |        261 |       mymodel.client.db1.db1
import time:       296 |    1901117 |     mymodel.client.db1
import time:   3223592 |    3223592 |       mymodel.client.db2.models
import time:       369 |        369 |       mymodel.client.db2.db2
import time:       309 |    3224269 |     mymodel.client.db2
import time:   4325926 |    4325926 |       mymodel.client.db3.models
import time:       111 |        111 |         mymodel.constant
import time:       717 |        828 |       mymodel.client.db3.db3
import time:       319 |    4327071 |     mymodel.client.db3
import time:       303 |   10304885 |   mymodel.client
import time:       310 |   10305195 | mymodel
```

Simple class without base and field/mapped:

```
...
import time:       399 |        399 |             mymodel.client.db2.db2
import time:       378 |    3833897 |           mymodel.client.db2
import time:    154845 |     154845 |             mymodel.client.db3.models
```

Full log:
[import2.log](https://github.com/sqlalchemy/sqlalchemy/files/11134501/import2.log)

**Comment by zzzeek** [comment, marked as answer]
3000 is a lot of tables.   I would normally say that's the first problem because it would be unusual to have a single program that contains enough different query objects to affect 3000 tables individually.

so if changing the number 3000 to something more reasonable like, 250, is not an option, you would want to break up the files into logical, non-dependent groups and then only import those sub-groups that you need.  there's no requirement to import all models up front.

**Comment by PaleNeutron** [comment]
I'll do that later, but can we speed up it on the other side. For example, make model dynamic and lazy load the slow code only when it is accessed.

**Reply by CaselIT** [reply]
It's not clear to me if you are using reflection on import, or you reflected the model once then wrote the python files with something like sqlacodegen

**Reply by PaleNeutron** [reply]
@CaselIT I reflected the model once  using sqlacodegen.

**Reply by CaselIT** [reply]
If you have all models defined in the same module I don't think there is much ability to speed up import.
You should consider splitting the module in multiple ones so you can import one only when needed.

That being said if profiling reveals some suboptimal operations in sqlalchemy we can improve them too

**Reply by zzzeek** [reply]
no model requires 3000 individual python classes
